File: 0039-combination-sum/0039-combination-sum.cpp
void fn(int ind, int target, vector<int>&arr, vector<vector<int>>&ds, vector<int>&candi)
{
    if(ind==candi.size()){
        if(target==0)
        {
            ds.push_back(arr);
        }
        return;
    }

    if(candi[ind]<=target){

        arr.push_back(candi[ind]);
        fn(ind,target-candi[ind],arr,ds,candi);
        arr.pop_back();
    }

    fn(ind+1,target,arr,ds,candi);
}
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> arr;
        vector<vector<int>> ds;

        fn(0, target, arr, ds, candidates);
        return ds;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private void fn(int ind, int[]candidates, List<Integer>ds, List<List<Integer>> ans, int target){
        if(candidates.length==ind){
            if(target==0){
                ans.add(new ArrayList <> (ds));
            }
            return ;
        }

        if(candidates[ind]<=target){

            ds.add(candidates[ind]);
            fn(ind, candidates, ds, ans, target-candidates[ind]);
            ds.remove(ds.size()-1);

        }

        fn(ind+1,candidates,ds,ans,target);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<List<Integer>> ans=new ArrayList<> (); 
        fn(0,candidates, new ArrayList<> (), ans, target);
        return ans;
    }
}